src/charm.py

```diff
@@ -0,0 +1,163 @@
+        self._container_name = self._service_name = "nrf"
+        self._container = self.unit.get_container(self._container_name)
+        self._database = DatabaseRequires(
+            self, relation_name="database", database_name=DATABASE_NAME
+        )
+        self.framework.observe(self.on.database_relation_joined, self._configure_pebble_layer)
+        self.framework.observe(self.on.nrf_pebble_ready, self._configure_pebble_layer)
+        self.framework.observe(self._database.on.database_created, self._on_database_created)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="sbi", port=NRF_SBI_PORT),
+            ],
+        )
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Handle database created event."""
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        self._write_config_file(
+            database_url=event.uris.split(",")[0],
+        )
+        self._configure_pebble_layer(event)
+
+    def _write_config_file(self, database_url: str) -> None:
+        """Writes config file to workload.
+
+        Args:
+            database_url: Database URL
+        """
+        jinja2_environment = Environment(loader=FileSystemLoader("src/templates/"))
+        template = jinja2_environment.get_template("nrfcfg.yaml.j2")
+        content = template.render(
+            database_name=DATABASE_NAME,
+            database_url=database_url,
+            nrf_sbi_port=NRF_SBI_PORT,
+        )
+        self._container.push(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}", source=content)
+        logger.info(f"Pushed {CONFIG_FILE_NAME} config file")
+
+    @property
+    def _config_file_is_written(self) -> bool:
+        """Returns whether the config file was written to the workload container."""
+        if not self._container.exists(f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
+            logger.info(f"Config file is not written: {CONFIG_FILE_NAME}")
+            return False
+        logger.info("Config file is written")
+        return True
+
+    def _configure_pebble_layer(
+        self, event: Union[PebbleReadyEvent, RelationCreatedEvent, DatabaseCreatedEvent]
+    ) -> None:
+        """Adds pebble layer and manages Juju unit status.
+
+        Args:
+            event: Juju event
+        """
+        if not self._database_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for database relation to be created")
+            return
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        if not self._config_file_is_written:
+            self.unit.status = WaitingStatus("Waiting for config file to be written")
+            return
```

Comment:
Shouldn't we `event.defer()` here?

Reply:
Not necessarily. Config file is created after DB is created and DB creation will trigger this whole function again anyway.

Reply:
No because on database created (where the config file is created), this same `_configure_pebble_layer` method will be called

Reply:
My bad.